### PR TITLE
add example for querying SHARP data and constructing a Map

### LINKED ISSUE #6663: Add an example gallery entry for querying SHARP data and constructing a Map
The SHARP data series is frequently used, but we do not have an example showing how to query the data or how to construct a Map for the data.

~~Currently, our `JSOCClient` does not support this, but~~ it can be done through `drms` directly, e.g. adapting the [example here](https://docs.sunpy.org/projects/drms/en/stable/generated/gallery/plot_synoptic_mr.html)

```python
import astropy.time
import drms

c = drms.Client()
sharp_num = 377
ts = astropy.time.Time('2011-02-10 22:00:00',scale='tai')
ds = f'hmi.sharp_cea_720s[{sharp_num}][{ts.tai.isot}_TAI/10h@1h]'
results, filenames = c.query(ds, key=drms.const.all, seg=['Bp','Br','Bt'])
```

Then, to download the first file for the phi component of the field ('Bp'),

```python
import astropy.io.fits

data = astropy.io.fits.getdata(f'http://jsoc.stanford.edu{filenames["Bp"][0]}')
header = dict(results.iloc[0])
m = sunpy.map.Map(data, header)
m.peek()
```

image

It may be worth looking into adding support through our JSOC client for this kind of query as well as adding support for these types of maps, either through a standalone map class or through the existing HMIMap class.

### PR DESCRIPTION
**Pull request by exitflynn**
## PR Description
Adds an example gallery entry for querying SHARP data and then constructing a Map. Closes #6663.

### CODE REVIEW

File: examples/acquiring_data/working_with_SHARP_data.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+###################################################################
+# We plug all these attributes into Fido to conduct the search.
+# Once the search is done we can use Fido to fetch the data.
+
+results = Fido.search(a.Time(tstart, tend),
+                      a.Sample(sampling_rate),
+                      a.jsoc.Series(series),
+                      a.jsoc.PrimeKey(primekey_label, primekey_value),
+                      a.jsoc.Notify(notify_email),
+                      a.jsoc.Segment(segment))
```
Comment (nabobalis) [inline examples/acquiring_data/working_with_SHARP_data.py:44]:
```suggestion
result = Fido.search(a.Time("2023-02-01 21:00:00", "2023-02-01 22:30:00"),
                      a.Sample(1*u.hour),
                      a.jsoc.Series("hmi.sharp_cea_720s"),
                      a.jsoc.PrimeKey("HARPNUM", 7871),
                      a.jsoc.Notify(notify_email),
                      a.jsoc.Segment("Bp"))
```
You need to use the notify email system we use from other JSOC examples.

File: examples/acquiring_data/working_with_SHARP_data.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
```
Comment (nabobalis) [inline examples/acquiring_data/working_with_SHARP_data.py:45]: # Then fetch the file.

File: examples/acquiring_data/working_with_SHARP_data.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+files = Fido.fetch(results)
```
Comment (nabobalis) [inline examples/acquiring_data/working_with_SHARP_data.py:46]:
```suggestion
file = Fido.fetch(result)
```

File: examples/acquiring_data/working_with_SHARP_data.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+###################################################################
+# Once the data has been queried, we can construct a map with it
+# using ``sunpy.map.Map()``
+
+sharp_bp = sunpy.map.Map(files)
```
Comment (nabobalis) [inline examples/acquiring_data/working_with_SHARP_data.py:52]:
```suggestion
# Now we have the file, we will construct a `sunpy.map.Map`.

sharp_map = sunpy.map.Map(file)
sharp_map.plot()

plt.show()
```

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import sunpy.map
+from sunpy.net import Fido
+from sunpy.net import attrs as a
+
+###################################################################
+# To get access to SHARP data, we will need to query the [JSOC/Stanford](http://jsoc.stanford.edu/).
+# We will use `Fido <sunpy.net.fido_factory.UnifiedDownloaderFactory>`
+# and make use of the JSOC attributes that allow us to query the JSOC.
+
+jsoc_email = os.environ["JSOC_EMAIL"]
```
Comment (nabobalis) [inline examples/acquiring_data/querying_and_loading_SHARP_data.py:21]:
```suggestion
# You will need to replace this with your own registered email address:
# http://jsoc.stanford.edu/ajax/register_email.html
jsoc_email = os.environ["JSOC_EMAIL"]
```

**Comment by nabobalis** [comment]
I am not sure if the example runs, can you check that I didn't break it?

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import os
+
+import astropy.units as u
+
+import sunpy.map
+from sunpy.net import Fido
+from sunpy.net import attrs as a
+
+###################################################################
+# To get access to SHARP data, we will need to query the [JSOC/Stanford](http://jsoc.stanford.edu/).
```
Comment (wtbarnes) [inline examples/acquiring_data/querying_and_loading_SHARP_data.py:17]:
```suggestion
# To search for SHARP data, we will need to query the [JSOC/Stanford](http://jsoc.stanford.edu/).
```

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+# We will use `Fido <sunpy.net.fido_factory.UnifiedDownloaderFactory>`
```
Comment (wtbarnes) [inline examples/acquiring_data/querying_and_loading_SHARP_data.py:18]: Is this actually what we want to link to in our API docs?

Comment (nabobalis) [reply examples/acquiring_data/querying_and_loading_SHARP_data.py:18]: I thought this is how we link Fido, but is there another way?

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+jsoc_email = os.environ["JSOC_EMAIL"]
+
+result = Fido.search(a.Time("2023-02-01 21:00:00", "2023-02-01 22:30:00"),
+                     a.Sample(1*u.hour),
+                     a.jsoc.Series("hmi.sharp_cea_720s"),
+                     a.jsoc.PrimeKey("HARPNUM", 7871),
+                     a.jsoc.Notify(jsoc_email),
+                     a.jsoc.Segment("Bp"))
+# Then fetch the file.
+file = Fido.fetch(result)
```
Comment (wtbarnes) [inline examples/acquiring_data/querying_and_loading_SHARP_data.py:30]:
```suggestion
print(result)

###################################################################
# Next, we can download the file

file = Fido.fetch(result)
```

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+###################################################################
+# Now we have the file, we will construct a `sunpy.map.Map`.
```
Comment (wtbarnes) [inline examples/acquiring_data/querying_and_loading_SHARP_data.py:33]:
```suggestion
# Now that we have the file, we can construct a `sunpy.map.Map` and plot it.
```

**Review by wtbarnes — CHANGES_REQUESTED** [review]
A few minor changes. The RTD preview has not rendered yet either and I'd like to take a brief look at that before approving.

File: examples/acquiring_data/querying_and_loading_SHARP_data.py (hunk 7 above)
Comment (nabobalis) [reply examples/acquiring_data/querying_and_loading_SHARP_data.py:18]: `sunpy.net.fido_factory.Fido` should work

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+import astropy.units as u
+
+import sunpy.map
+from sunpy.net import Fido
+from sunpy.net import attrs as a
+
+###################################################################
+# To search for SHARP data, we will need to query the [JSOC/Stanford](http://jsoc.stanford.edu/).
+# We will use `Fido <sunpy.net.fido_factory.Fido>`
```
Comment (nabobalis) [inline examples/acquiring_data/querying_and_loading_SHARP_data.py:18]: This in fact did not work: `/home/runner/work/sunpy/sunpy/docs/generated/gallery/acquiring_data/querying_and_loading_SHARP_data.rst:42: WARNING: py:obj reference target not found: sunpy.net.fido_factory.Fido`

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import astropy.units as u
+
+import sunpy.map
+from sunpy.net import Fido
+from sunpy.net import attrs as a
+
+###################################################################
+# To get access to SHARP data, we will need to query the [JSOC/Stanford](http://jsoc.stanford.edu/).
+# We will use `Fido <sunpy.net.fido_factory.UnifiedDownloaderFactory>`
+# and make use of the JSOC attributes that allow us to query the JSOC.
```
[the remark this answers is not in the document]
Comment (exitflynn) [reply examples/acquiring_data/querying_and_loading_SHARP_data.py:19]: i'm afraid i'm having a bit of trouble understanding this. I'm read up the documentation and am familiar with intersphinx linking to things in the codebase but unsure about URL's, would just something like \``https://docs.sunpy.org/en/latest/reference/net.html#module-sunpy.net.jsoc`` do it?

File: examples/acquiring_data/querying_and_loading_SHARP_data.py (hunk 10 above)
Comment (exitflynn) [reply examples/acquiring_data/querying_and_loading_SHARP_data.py:18]: hi, i'm unaware how one checks this. I could revert it back to `sunpy.net.fido_factory.UnifiedDownloaderFactory` which was working before but would like to check if `sunpt.net.Fido` would work, could you point me to what documentation will be helpful here?

Comment (nabobalis) [reply examples/acquiring_data/querying_and_loading_SHARP_data.py:18]: Go back to `sunpy.net.fido_factory.UnifiedDownloaderFactory` for now.

Comment (exitflynn) [reply examples/acquiring_data/querying_and_loading_SHARP_data.py:18]: sure, thanks!

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+import sunpy.map
+from sunpy.net import Fido
+from sunpy.net import attrs as a
+
+###################################################################
+# To search for SHARP data, we will need to query the `JSOC <http://jsoc.stanford.edu/>`__.
+# We will use `Fido <sunpy.net.fido_factory.UnifiedDownloaderFactory>`
+# and make use of the JSOC attributes that allow us to query the JSOC.
+
```
Comment (nabobalis) [inline examples/acquiring_data/querying_and_loading_SHARP_data.py:20]:
```suggestion
#
```

**Comment by exitflynn** [comment]
i think there mustn't be anymore changes left to make now?

**Comment by nabobalis** [comment]
Have you checked if the example works and renders nicely on RTD?

**Comment by exitflynn** [comment]
i built the docs by myself using `sphinx-build` and there were indeed a minor problem. all fixed now, everything looks good to me!

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+                     a.jsoc.Segment("Bp"))
+print(result)
+
+###################################################################
+# Next, we can download the file
+
+file = Fido.fetch(result)
+
+###################################################################
+# Now that we have the file, we can construct a `sunpy.map.Map` and plot it.
```
Comment (wtbarnes) [inline examples/acquiring_data/querying_and_loading_SHARP_data.py:42]:
```suggestion
# Now that we have the file, we can construct a `~sunpy.map.Map` and plot it.
```

**Comment by wtbarnes** [comment]
This still needs a changelog.

File: examples/acquiring_data/querying_and_loading_SHARP_data.py (hunk 13 above)
Comment (exitflynn) [reply examples/acquiring_data/querying_and_loading_SHARP_data.py:42]: thanks for catching that :grimacing: :see_no_evil:

File: changelog/6757.doc.rst [hunk 14]
```diff
@@ -0,0 +1 @@
+Added an example gallery entry for querying SHARP data and loading it into a `Map`.
```
[the remark this answers is not in the document]
Comment (wtbarnes) [reply changelog/6757.doc.rst:1]: The incomplete reference was what was causing the doc build failure.

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+###################################################################
+# Next, we can download the file
+
+file = Fido.fetch(result)
+
+###################################################################
+# Now that we have the file, we can construct a `~sunpy.map.Map` and plot it.
+
+sharp_map = sunpy.map.Map(file)
+sharp_map.plot(vmin=-1500, vmax=1500)
```
Comment (wtbarnes) [inline examples/acquiring_data/querying_and_loading_SHARP_data.py:45]:
```suggestion
plt.figure()
sharp_map.plot(vmin=-1500, vmax=1500)
```

per dev docs for writing gallery examples: https://docs.sunpy.org/en/latest/dev_guide/contents/example_gallery.html#contribution-guidelines. I thought this was already included for some reason...

Comment (nabobalis) [reply examples/acquiring_data/querying_and_loading_SHARP_data.py:45]: We never updated it?! I will get on that.

File: examples/acquiring_data/querying_and_loading_SHARP_data.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+result = Fido.search(a.Time("2011-03-09 23:20:00", "2011-03-09 23:30:00"),
+                     a.Sample(1*u.hour),
+                     a.jsoc.Series("hmi.sharp_cea_720s"),
+                     a.jsoc.PrimeKey("HARPNUM", 401),
+                     a.jsoc.Notify(jsoc_email),
+                     a.jsoc.Segment("Bp"))
+print(result)
+
+###################################################################
+# Next, we can download the file
```
Comment (nabobalis) [inline examples/acquiring_data/querying_and_loading_SHARP_data.py:37]:
```suggestion
# Next, we can download the file.
```

File: changelog/6757.doc.rst [hunk 17]
```diff
@@ -0,0 +1 @@
+Added an example gallery entry for querying SHARP data and loading it into a `~sunpy.map.Map`.
```
Comment (nabobalis) [inline changelog/6757.doc.rst:1]:
```suggestion
Added an example gallery (:ref:`sphx_glr_generated_gallery_acquiring_data_querying_and_loading_SHARP_data.py
`) for querying SHARP data and loading it into a `~sunpy.map.Map`.
```

**Comment by nabobalis** [comment]
I will merge after I glance over the RTD version.

**Comment by nabobalis** [comment]
Thanks @exitflynn for the example

**Comment by exitflynn** [comment]
thank you so much for the reviews as well, Nabil and Will! :cowboy_hat_face:

**Pull request merged**